File: Mandelbrot.Lib.CSharp/Mandelbrot.cs
namespace Mandelbrot.Lib.CSharp
{
    public class Mandelbrot
    {
        private readonly int _maxIterations;
        private readonly float _limit;

        public Mandelbrot(int maxIterations, float limit)
        {
            _maxIterations = maxIterations;
            _limit = limit;
        }

        /// <summary>
        /// Whether the recursive calculation Z = (Z * Z) + C remains within bounds 
        ///  during iterations starting with Z = 0.
        /// </summary>
        /// <param name="c">ComplexNumber value of C.</param>
        /// <returns>Result type instance.</returns>
        public Result IsDivergent(ComplexNumber c)
        {
            // Iteration
            var i = 0;
            // Accumulator
            var z = ComplexNumber.Zero;

            while (i < _maxIterations)
            {
                if (ComplexNumber.Abs(z) > _limit)
                {
                    return new Result(true, i);
                }

                z = ComplexNumber.Square(z) + c;
            }

            return new Result(false, -1);
        }
    }

    public class Result
    {
        public bool IsDivergent { get; }
        public int DivergedAt { get; }

        public Result(bool isDivergent, int divergedAt)
        {
            IsDivergent = isDivergent;
            DivergedAt = divergedAt;
        }
    }
}

File: Mandelbrot.Lib/Mandelbrot.fs
namespace Mandelbrot.Lib.FSharp

module Mandelbrot =

    type Result = 
        | Bounded
        | Diverged of int

    /// Whether the recursive calculation Z = (Z * Z) + C remains less than a limit
    ///  during iterations starting with Z = 0.
    let isDivergent (maxIterations: int) (limit: float) (c: Complex.Number) = 
        let rec loop i z = 
            if i > maxIterations then Bounded
            else if Complex.abs z > limit then Diverged i
            else loop (i + 1) (Complex.square z |> Complex.add c)

        loop 0 Complex.zero